File: neetcode_150/bestTimetoBuySellStock.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
public class bestTimetoBuySellStock {
    public static void main(String[] args) {

    }
    public int maxProfit(int[] prices) {
        if(prices.length<=1){
            return 0;
        }
        int buy = prices[0];
        int sell = prices[1];
        int profit = 0;
        for(int i =0; i<prices.length; i++){
            if(prices[i]<buy){
                buy = prices[i];
                sell = 0;
            }
            else if(prices[i]>sell){
                sell = prices[i];
            }
            profit = Math.max(profit, sell-buy);
        }
        return profit;
    }
}

File: neetcode_150/bestTimeToBuySellStock.py
import sys
from typing import List


def maxProfit(self, prices: List[int]) -> int:
    if len(prices) <= 1:
        return 0
    buy = prices[0]
    sell = prices[1]
    result = sell - buy
    for i in range(1, len(prices) - 1):
        if buy > prices[i]:
            buy = prices[i]
            sell = 0
        if (sell < prices[i + 1]):
            sell = prices[i + 1]
        result = max(result, sell - buy)
    if result < 0:
        result = 0
    return result


